test/e2e/settings/disable_catch_all.go

```diff
@@ -0,0 +1,94 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package settings
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	extensions "k8s.io/api/extensions/v1beta1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	"k8s.io/ingress-nginx/test/e2e/framework"
+)
+
+const (
+	waitForLuaSync = 5 * time.Second
+)
+
+var _ = framework.IngressNginxDescribe("Disabled catch-all", func() {
+	f := framework.NewDefaultFramework("disabled-catch-all")
+
+	BeforeEach(func() {
+		f.NewEchoDeployment()
+	})
+
+	AfterEach(func() {
+	})
+
+	It("should ignore catch all Ingress", func() {
+
+		name := "foo"
+		ingKey := fmt.Sprintf("%v/%v", f.IngressController.Namespace, name)
+
+		ing := framework.NewSingleCatchAllIngress(name, f.IngressController.Namespace, "http-svc", 80, nil)
+		f.EnsureIngress(ing)
+
+		time.Sleep(waitForLuaSync)
+
+		logs, err := f.NginxLogs()
+		Expect(err).ToNot(HaveOccurred())
+		Expect(logs).To(ContainSubstring(`ignoring add for catch-all ingress ` + ingKey + ` because of --disable-catch-all`))
+	})
+
+	It("should delete Ingress updated to catch-all", func() {
+
+		host := "foo"
+		ingKey := fmt.Sprintf("%v/%v", f.IngressController.Namespace, host)
+
+		ing := framework.NewSingleIngress(host, "/", host, f.IngressController.Namespace, "http-svc", 80, nil)
+		f.EnsureIngress(ing)
+
+		f.WaitForNginxServer(host,
+			func(server string) bool {
+				return strings.Contains(server, "server_name foo")
+			})
+
+		err := framework.UpdateIngress(f.KubeClientSet, f.IngressController.Namespace, host, func(ingress *extensions.Ingress) error {
+			ingress.Spec.Rules = nil
+			ingress.Spec.Backend = &extensions.IngressBackend{
+				ServiceName: "http-svc",
+				ServicePort: intstr.FromInt(80),
+			}
+			return nil
+		})
+		Expect(err).ToNot(HaveOccurred())
+
+		time.Sleep(waitForLuaSync)
+
+		f.WaitForNginxConfiguration(func(cfg string) bool {
+			return !strings.Contains(cfg, "server_name foo")
+		})
```

Comment:
IMO asserting using HTTP request is a lot better. It represents "end-to-endness" better.

Reply:
I actually cannot do that.

Moving the DisableCatchAll option to the store, make the default catch-all server still present.
So, when updating my ingress, request get routed to this catch-all server and answer a 200.

Are we good with having the default catch-all server still present even with `--disable-catch-all` ?

Reply:
default catch-all server returns 404

Reply:
Alright, I've found out why my GET query returns a 200 instead of a 404.

It seems that when an `Ingress` is updated, this https://github.com/Shopify/ingress/pull/184/files#diff-ece58b3bebec02a6d3173c8d8099d298R386 does not actually delete the `Ingress` from the controller.
The Ingress gets updated and then merge into the default backend.

Also, this showed me that we have a similar issue with the `kubernetes.io/ingress.class` annotation. On the same principle, if you have a running, valid `Ingress`, and you update its `ingress.class` to something invalid. It does not remove the old `Ingress` there https://github.com/kubernetes/ingress-nginx/blob/master/internal/ingress/controller/store/store.go#L365

Is that done on purpose or actually a bug?